src/Microsoft.ML.Data/Evaluators/Metrics/ConfusionMatrix.cs

```diff
@@ -0,0 +1,49 @@
+﻿using System;
+using System.Collections.Generic;
+using Microsoft.ML.Runtime;
+
+namespace Microsoft.ML.Data
+{
+    public sealed class ConfusionMatrix
+    {
+        public double[] PrecisionSums { get; }
+        public double[] RecallSums { get; }
+        public ReadOnlyMemory<char>[] Labels => PredictedLabelNames.ToArray(); //sefilipi: return as string[]
```

Comment:
>ReadOnlyMemory [](start = 15, length = 14)

Remember, as before, that we provide access to the columns, but do not copy out the annotations data. Remember the discussion around having names.

Remember also, as we've found in #3101 that the labels may not even be text, or even exist.

Reply:
I looked a bit on how is the ConfusionMatrix idv build, and it looks like an indicator for the predictedLabels will always be there, and it will be of the ReadOnlyMemory<char> type. 

[https://github.com/dotnet/machinelearning/blob/master/src/Microsoft.ML.Data/Evaluators/MulticlassClassificationEvaluator.cs#L222](https://github.com/dotnet/machinelearning/blob/master/src/Microsoft.ML.Data/Evaluators/MulticlassClassificationEvaluator.cs#L222)
[https://github.com/dotnet/machinelearning/blob/master/src/Microsoft.ML.Data/Evaluators/BinaryClassifierEvaluator.cs#L361](https://github.com/dotnet/machinelearning/blob/master/src/Microsoft.ML.Data/Evaluators/BinaryClassifierEvaluator.cs#L361)

The metrics aggregators have the class names in the ctor arguments, and if the class names are missing, those get generated from the size of the Score column. 

[https://github.com/dotnet/machinelearning/blob/master/src/Microsoft.ML.Data/Evaluators/MulticlassClassificationEvaluator.cs#L114](https://github.com/dotnet/machinelearning/blob/master/src/Microsoft.ML.Data/Evaluators/MulticlassClassificationEvaluator.cs#L114)
[https://github.com/dotnet/machinelearning/blob/master/src/Microsoft.ML.Data/Evaluators/BinaryClassifierEvaluator.cs#L184](https://github.com/dotnet/machinelearning/blob/master/src/Microsoft.ML.Data/Evaluators/BinaryClassifierEvaluator.cs#L184)

So given that, i left the labels as  ReadOnlyMemory<char> instead of providing the annotations of the Count column, and just renamed it to PredictedClassIndicators, since they are not going to be the class names all the time. 

Let me know if I am missing something else. 

---
In reply to: [273615151](https://github.com/dotnet/machinelearning/pull/3250#discussion_r273615151) [](ancestors = 273615151)

Reply:
> I looked a bit on how is the ConfusionMatrix idv build, and it looks like an indicator for the predictedLabels will always be there, and it will be of the `ReadOnlyMemory<char>` type.

Yes, but AFAIK that is mostly because it existed in order to provide a string to write to console, for the console application, back when this was purely a command line tool and not an API. So it has to have *some* string description, that doesn't mean that it's a good string description. I think in fact it'll just render the number as a string in case there is none. So... *not* something we want to expose as part of the API!